### PR TITLE
feat(lookml): Use Looker API to get fields of a View

### PR DESCRIPTION
**Pull request by askumar27**
# 📋 Summary

**Enhanced LookML ingestion with field splitting, parallel processing, and resilient lineage extraction for large views**

This PR introduces robust handling for Looker views with large field sets by implementing field splitting, parallel processing, and individual field fallback mechanisms. These improvements prevent SQL parsing failures, improve performance, and maximize lineage extraction even when some fields fail.

---

## 🎯 Motivation

**Problem Statement:**
- Large Looker views with hundreds of fields were causing SQL parsing failures when querying the Looker API for lineage
- Single large API queries with many fields resulted in complex SQL that exceeded parsing capabilities
- Complete failure of lineage extraction when any part of the query failed, leading to zero lineage for large views
- Inefficient API usage when multiple views could be served by the same explore

**Business Impact:**
- Users with large Looker views (common in enterprise environments) were unable to extract lineage
- Missing lineage information impacts data governance, impact analysis, and compliance reporting
- Performance bottlenecks when processing large field sets sequentially

**Solution:**
This PR implements a multi-layered resilience strategy:
1. **Field Splitting**: Automatically splits large field sets into manageable chunks
2. **Parallel Processing**: Processes chunks concurrently to improve performance
3. **Individual Field Fallback**: When chunks fail, processes fields individually to maximize lineage
4. **View-to-Explore Optimization**: Minimizes API calls by intelligently mapping views to explores

---

### New Features

**1. Field Splitting for Large Views**
- Automatically splits views with >100 fields (configurable) into multiple API calls
- Each chunk is processed independently, preventing one failure from affecting others
- Results are combined to provide comprehensive lineage

**2. Parallel Processing**
- Uses `ThreadPoolExecutor` to process field chunks concurrently
- Configurable worker count (default: 10, max: 100) for performance tuning
- Significantly reduces processing time for large views

**3. Individual Field Fallback**
- When a chunk fails, automatically attempts to process each field individually
- Maximizes lineage extraction by isolating problematic fields
- Provides detailed reporting on which fields cause issues

**4. View-to-Explore Optimization**
- Implements greedy algorithm to minimize API calls
- Groups views that can be served by the same explore
- Logs optimization efficiency metrics

**5. Partial Lineage Support**
- Allows partial lineage results even when some chunks fail
- Configurable via `allow_partial_lineage_results` flag
- Ensures some lineage is available rather than complete failure

---

### Testing Approach

**Unit Tests:**
- Comprehensive coverage of all new methods
- Mock-based testing for Looker API interactions
- Edge case testing (empty fields, all failures, single field, etc.)
- Configuration validation testing

**Integration Tests:**
- End-to-end pipeline testing with mocked Looker API
- Performance testing for parallel processing
- Error scenario testing (chunk failures, parsing errors)
- Optimization algorithm validation

### Key Test Scenarios

1. **Field Splitting Trigger**: Verifies splitting occurs when threshold exceeded
2. **Result Combination**: Ensures results from multiple chunks are correctly merged
3. **Individual Field Fallback**: Tests fallback when chunks fail
4. **Partial Lineage**: Validates partial results are returned when enabled
5. **Parallel Processing**: Confirms concurrent execution with multiple workers
6. **Optimization**: Verifies view-to-explore mapping reduces API calls
7. **Error Handling**: Tests graceful degradation on various failure modes

### Test Coverage

- **Unit Tests**: 10 new test methods covering all critical paths
- **Integration Tests**: 6 integration tests for end-to-end validation
- **Edge Cases**: Handles empty fields, single fields, all failures, parsing errors

---

### Affected Components

**Primary:**
- `src/datahub/ingestion/source/looker/view_upstream.py` - Core lineage extraction logic
- `src/datahub/ingestion/source/looker/lookml_config.py` - Configuration options
- `src/datahub/ingestion/source/looker/lookml_source.py` - Source orchestration

**Secondary:**
- `src/datahub/ingestion/source/looker/looker_lib_wrapper.py` - API wrapper enhancements
- Test files for unit and integration testing

### Breaking Changes

**None** - All changes are backward compatible:
- New configuration options have sensible defaults
- Existing behavior preserved for views below threshold
- All new features are opt-in via configuration

### Performance Impact

**Positive:**
- **Parallel Processing**: Significantly reduces processing time for large views (up to 10x faster with default workers)
- **API Optimization**: Reduces API calls through view-to-explore optimization
- **Caching**: Explore field information cached to reduce redundant API calls

**Considerations:**
- **Memory**: Parallel processing increases memory usage (configurable via worker count)
- **API Rate Limits**: More concurrent requests may approach rate limits (mitigated by configurable workers)

### Risk Level

**Low to Medium**

**Low Risk Factors:**
- Backward compatible with sensible defaults
- Comprehensive test coverage
- Graceful error handling and fallback mechanisms
- Configurable thresholds and limits

**Medium Risk Factors:**
- Parallel processing introduces concurrency complexity
- Large field sets may still encounter edge cases
- API rate limiting concerns with high worker counts

**Mitigation:**
- Default worker count (10) is conservative
- Maximum worker limit (100) prevents resource exhaustion
- Individual field fallback ensures maximum lineage extraction
- Partial lineage support prevents complete failures

---

### Configuration Changes

**New Optional Configuration Parameters:**
```yaml
source:
  type: lookml
  config:
    # New optional parameters (all have sensible defaults)
    field_threshold_for_splitting: 100  # Trigger splitting above this count
    allow_partial_lineage_results: true  # Allow partial results on errors
    enable_individual_field_fallback: true  # Process fields individually on chunk failure
    max_workers_for_parallel_processing: 10  # Parallel processing workers (1-100)
```

**No Breaking Changes:**
- All parameters are optional with defaults
- Existing configurations continue to work without modification
- Default behavior matches previous implementation for small views

---

### Trade-offs Considered

1. **Chunk Size vs. API Calls**: Larger chunks = fewer calls but higher failure risk. Current approach balances both.

2. **Parallelism vs. Resource Usage**: More workers = faster processing but higher memory/API load. Configurable to allow tuning.

3. **Partial Results vs. Strict Validation**: Allowing partial results improves resilience but may mask issues. Configurable to allow strict mode.

---

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/looker/lookml_source.py [hunk 1]
```diff
@@ -789,8 +793,10 @@ def get_internal_workunits(self) -> Iterable[Union[MetadataWorkUnit, Entity]]:
                         for view_name in explore.upstream_views:
                             if self.source_config.emit_reachable_views_only:
                                 explore_reachable_views.add(view_name.include)
-                            # Build view to explore mapping for API-based view lineage
-                            view_to_explore_map[view_name.include] = explore.name
+
+                            # Build view-to-explores mapping efficiently
```
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/ingestion/source/looker/lookml_source.py]: what does efficiently mean here?

File: metadata-ingestion/src/datahub/ingestion/source/looker/lookml_source.py [hunk 2]
```diff
@@ -1040,5 +1046,63 @@ def report_skipped_unreachable_views(
...
+            explore: len(views) for explore, views in explore_to_views.items()
+        }
+
+        # Build view-to-explore mapping using dynamic programming approach
+        view_to_explore: Dict[str, str] = {}
+
+        # For each view, find the explore with maximum size that contains it
+        for view_name in view_to_explores:
+            # Get all explores that contain this view from pre-built mapping
+            candidate_explores = view_to_explores[view_name]
```
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/ingestion/source/looker/lookml_source.py]: isn't this the same?

```suggestion
        # For each view, find the explore with maximum size that contains it
        for view_name, candidate_explores in view_to_explores.items():
```

or we cannot do that because we are updating `view_to_explore[view_name]` in the same iteration?

**Comment by sgomezvillamor** [comment]
> * We use `lkml` parser to get View attributes, and this has limitation of when Views are very complex
> * Use Looker APIs to get fields of a View, this API need to be called via the Explore name.

Are we adding a fallback strategy? or do we replace with new strategy?
If the second, that could lead to some breaking change. If so, how can we address or mitigate it?

**Comment by sgomezvillamor** [comment]
Looker/LookML ingestion is becoming a very complex code.
That's noted in the amount of debug logging required. 

Given the amount of detail we require for troubleshooting, I wonder if there is python tooling that we can use to trace execution and no require explicit logger debug lines. 🤔 

Not a blocker, just thinking loudly.

File: metadata-ingestion/src/datahub/ingestion/source/looker/lookml_source.py (hunk 1 above)
Comment (askumar27) [reply metadata-ingestion/src/datahub/ingestion/source/looker/lookml_source.py]: This comment is misplaced when I was moving around the function, it refers to `_optimize_views_by_common_explore`. Will update, here we are just populating the data

File: metadata-ingestion/src/datahub/ingestion/source/looker/lookml_source.py (hunk 2 above)
Comment (askumar27) [reply metadata-ingestion/src/datahub/ingestion/source/looker/lookml_source.py]: Let me check - they should be equivalent.

**Comment by askumar27** [comment]
> > * We use `lkml` parser to get View attributes, and this has limitation of when Views are very complex
> > * Use Looker APIs to get fields of a View, this API need to be called via the Explore name.
> 
> Are we adding a fallback strategy? or do we replace with new strategy? If the second, that could lead to some breaking change. If so, how can we address or mitigate it?

There is a fallback strategy in place if there are any issues with Looker APIs to get the fields, it falls back to view_context based solution (current). There should not be any backward incompatibility issues here with either of the strategy. Worst case scenarios is no CLL or partial CLL (same as today)

> - Automatic fallback to view_context based fields extraction with better error handling
> - Best effort fields extraction - continue on failure (atleast one field is required) instead of falling back to different upstream strategy

**Comment by askumar27** [comment]
> Looker/LookML ingestion is becoming a very complex code. That's noted in the amount of debug logging required.
> 
> Given the amount of detail we require for troubleshooting, I wonder if there is python tooling that we can use to trace execution and no require explicit logger debug lines. 🤔
> 
> Not a blocker, just thinking loudly.

This is a great suggestion for `tracing`, we should certainly explore options. 
Regarding the debug logs, Sorry I forgot to mention in the PR - this is also a dev build to be used for a customer debug. I will then trim out the debug logs before merging.

File: metadata-ingestion/src/datahub/ingestion/source/looker/lookml_source.py (hunk 2 above)
Comment (askumar27) [reply metadata-ingestion/src/datahub/ingestion/source/looker/lookml_source.py]: Updated!
